commit 16730ec7af373e182b2ea5ca5a2f6973fca954b4
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Tue Apr 16 22:12:59 2024 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Tue Apr 16 22:12:59 2024 -0400

    [2.3] Improve dead client handling (#2501)
    
    * improve dead client handling
    
    * remove unused import
    
    * fix docstring
    
    ---------
    
    Co-authored-by: Yuan-Ting Hsieh (謝沅廷) <yuantingh@nvidia.com>
    Co-authored-by: Chester Chen <512707+chesterxgchen@users.noreply.github.com>
---
 nvflare/private/fed/server/server_engine.py | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 17be86c4..b06f5bb0 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -583,6 +583,19 @@ class ServerEngine(ServerEngineInternalSpec):
                 message=request,
             )
 
+    def notify_dead_job(self, job_id: str, client_name: str, reason: str):
+        shareable = Shareable()
+        shareable.set_header(ServerCommandKey.FL_CLIENT, client_name)
+        shareable.set_header(ServerCommandKey.REASON, reason)
+        self.send_command_to_child_runner_process(
+            job_id=job_id,
+            command_name=ServerCommandNames.HANDLE_DEAD_JOB,
+            command_data=shareable,
+            timeout=0.0,
+            optional=True,
+        )
+        self.logger.warning(f"notified SJ of dead-job: {job_id=}; {client_name=}; {reason=}")
+
     def send_command_to_child_runner_process(
         self, job_id: str, command_name: str, command_data, timeout=5.0, optional=False
     ):
@@ -594,7 +607,7 @@ class ServerEngine(ServerEngineInternalSpec):
                     targets=fqcn,
                     channel=CellChannel.SERVER_COMMAND,
                     topic=command_name,
-                    request=request,
+                    message=request,
                     optional=optional,
                 )
                 return None
